ydb/core/kqp/ut/tx/kqp_locks_tricky_ut.cpp

```diff
@@ -336,5 +337,251 @@ Y_UNIT_TEST_SUITE(KqpLocksTricky) {
+                    UNIT_ASSERT(evWrite->Record.GetMvccSnapshot().GetStep() == 0);
+                    UNIT_ASSERT(evWrite->Record.GetMvccSnapshot().GetTxId() == 0);
+                    hasWrite = true;
+                }
+
+                return TTestActorRuntime::EEventAction::PROCESS;
+            };
+
+            TDispatchOptions opts;
+            opts.FinalEvents.emplace_back([&](IEventHandle&) {
+                return hasWrite;
+            });
+
+            runtime.SetObserverFunc(grab);
+
+            auto future = kikimr.RunInThreadPool([&]{
+                auto txc = TTxControl::BeginTx(TTxSettings::SerializableRW()).CommitTx();
+                return session.ExecuteDataQuery(query, txc, execSettings).ExtractValueSync();
+            });
+
+            auto result = runtime.WaitFuture(future);
+            UNIT_ASSERT_VALUES_EQUAL_C(result.GetStatus(), EStatus::SUCCESS, result.GetIssues().ToString());    
+
+            
+            runtime.DispatchEvents(opts);
+            UNIT_ASSERT(hasWrite);    
+        }
+    }
+
+    Y_UNIT_TEST_TWIN(TestSnapshotWithDependentReads, UseSink) {
+        NKikimrConfig::TAppConfig appConfig;
+        appConfig.MutableTableServiceConfig()->SetEnableOltpSink(UseSink);
+
+        auto setting = NKikimrKqp::TKqpSetting();
+        TKikimrSettings settings;
+        settings.SetAppConfig(appConfig);
+        settings.SetUseRealThreads(false);
+        TKikimrRunner kikimr(settings);
+        auto db = kikimr.GetTableClient();
+        auto session = kikimr.RunCall([&] { return db.CreateSession().GetValueSync().GetSession(); });
+        auto upsertSession = kikimr.RunCall([&] { return db.CreateSession().GetValueSync().GetSession(); });
+
+        auto& runtime = *kikimr.GetTestServer().GetRuntime();
+        NYdb::NTable::TExecDataQuerySettings execSettings;
+        execSettings.CollectQueryStats(ECollectQueryStatsMode::Basic);
+
+        {
+                const TString upsertQuery(Q1_(R"(
+                    UPSERT INTO `/Root/KeyValue` (Key, Value) VALUES (1u, "One");
+                    UPSERT INTO `/Root/KeyValue2` (Key, Value) VALUES ("One", "expected");
+                )"));
+
+                auto upsertResult = kikimr.RunCall([&]{
+                    auto txc = TTxControl::BeginTx(TTxSettings::SerializableRW()).CommitTx();
+                    return upsertSession.ExecuteDataQuery(upsertQuery, txc, execSettings).ExtractValueSync();
+                });
+
+                UNIT_ASSERT_VALUES_EQUAL_C(upsertResult.GetStatus(), EStatus::SUCCESS, upsertResult.GetIssues().ToString());    
+        }
+        
+        {
+            const TString query(Q1_(R"(
+                $cnt1 = SELECT Value FROM `/Root/KeyValue` WHERE Key = 1u;
+                SELECT Ensure("ok", $cnt1="One", "first error");
+
+                $cnt2 = SELECT Value FROM `/Root/KeyValue2` WHERE Key = $cnt1;
+                SELECT Ensure("ok", $cnt2="expected", "second error");
```

Comment:
Without fix test fails here. 
First select reads old value and second select reads new value.